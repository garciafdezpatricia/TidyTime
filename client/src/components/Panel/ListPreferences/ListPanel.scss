@use "../../../../styles/variables" as *;

.list-preferences {
    padding: 1rem;
    display: flex;
    flex-direction: column;
    width: calc(100% - 2rem);
    p {
        font-size: .9rem;
    }   
}

.default-labels {
    font-size: .9rem;
    display: flex;
    flex-direction: column;
    margin-left: 1.5rem;
    margin-top: .6rem;
}

.labels-input {
    display: flex;
    flex-direction: column;
    width: fit-content;
    gap: .5rem;
    padding: .6rem;
    background-color: $background-button;
    border-radius: .4rem;
}

.individual-label {
    display: flex;
    gap: .5rem;
    align-items: center;

    input[type="text"] {
        width: 8rem;
        padding-left: .3rem;
        border-radius: .4rem;
        border: .1rem solid $light-grey;
        outline: none;
    }

    input[type="color"] {
        width: 1rem;
        height: 1rem;
        padding-left: .3rem;
        border-radius: 100%;
        border: .1rem solid transparent;
        outline: none;
        cursor: pointer;
    }
}

.delete-label-button {
    border: none;
    border-radius: .4rem;
    background-color: transparent;
    &:hover {
        color: $remove-red;
    }
}

.new-label{
    margin-top: .3rem;
    display: flex;
    width: fit-content;
    padding: .3rem;
    border: .1rem solid transparent;
    background-color: transparent;
    font-size: 1rem;
    &:hover {
        background-color: $primary-grey;
    }
}


.show-tasks-in-calendar {
    display: flex;
    gap: 1rem;
    font-size: .9rem;
    margin-bottom: 1rem;
    margin-left: 1.5rem;

    .show-buttons {
        display: flex;
        gap: .5rem;

        button {
            background-color: $background-button;
            border: none;
            padding-inline: .5rem;
        }

        .active-button {
            outline: .1rem solid $primary-blue;
        }
    }
}