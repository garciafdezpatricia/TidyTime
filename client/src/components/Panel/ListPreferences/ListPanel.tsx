// Copyright 2024 Patricia García Fernández.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { LuListPlus } from "react-icons/lu"
import { RiDeleteBin5Fill } from "react-icons/ri"
import { useTaskContext } from "../../Context/TaskContext"
import { useTranslation } from 'react-i18next';


export default function ListPanel() {
    const { t } = useTranslation();
    const { labels, setLabels, showTasksInCalendar, setshowTasksInCalendar } = useTaskContext();


    const handleLabelRename = (value:string, index:number) => {
        // @ts-ignore
        let result = [...labels];
        let updatedColumn = result[index];
        updatedColumn.name = value;
        result = [...result.slice(0, index),
            updatedColumn,
            ...result.slice(index + 1)
        ];
        setLabels(result);
    }

    const handleLabelDeletion = (index:number) => {
        let result = labels?.filter((label, labelIndex) => index !== labelIndex);
        setLabels(result);
    }

    const addNewInput = () => {
        // @ts-ignore
        let result = [...labels, {name: `New label ${labels.length + 1}`, color: "#000000"}];
        setLabels(result);
    }

    const handleLabelColor = (color:string, index:number) => {
        // @ts-ignore
        let result = [...labels];
        let updatedColumn = result[index];
        updatedColumn.color = color;
        result = [...result.slice(0, index),
            updatedColumn,
            ...result.slice(index + 1)
        ];
        setLabels(result);
    }

    return (
        <article className="list-preferences">
            <h3>{t('preferences.list.title')}</h3>
            <hr></hr>
            <p>{t('preferences.list.desc')}</p>
            <section className="default-labels">
                {
                    labels !== undefined 
                    ?
                    <div className="labels-input">
                    {
                        labels.length > 0 ?
                        labels.map((column, index) => {
                            return (
                                <div key={index} className="individual-label">
                                    <input style={{backgroundColor: column.color}}
                                        title={column.color ?? 'color'}
                                        type="color"
                                        defaultValue={column.color}
                                        onChange={(e) => handleLabelColor(e.target.value, index)}
                                    />
                                    <input 
                                        title={column.name ?? "tag"}
                                        type="text" 
                                        defaultValue={column.name}
                                        onChange={(e) => handleLabelRename(e.target.value, index)}
                                    />
                                    <p  
                                        onClick={() => handleLabelDeletion(index)}
                                        className="delete-label-button">
                                        <RiDeleteBin5Fill />
                                    </p>
                                </div>
                            )
                        })
                        : <div style={{paddingInline: ".5rem"}}><p style={{fontSize: ".7rem"}}>Create a new tag!</p></div>
                    }
                    </div>
                    :
                    <div style={{display: "flex", alignItems: "center", gap: ".5rem"}}>
                        <p>Loading...</p>
                        <div className="loader"></div>
                    </div>
                }
                <div 
                    onClick={addNewInput}
                    className="new-label">
                    <LuListPlus />
                </div>
            </section>
            <section className="show-tasks-in-calendar">
                <p>{t('preferences.list.showTasks')}</p>
                <div className="show-buttons">
                    <button
                        className={showTasksInCalendar ? "active-button" : ""}
                        onClick={() => setshowTasksInCalendar(true)}>
                        {t('preferences.list.showYes')}
                    </button>
                    <button 
                        className={showTasksInCalendar ? "" : "active-button"}
                        onClick={() => setshowTasksInCalendar(false)}>
                        {t('preferences.list.showNo')}
                    </button>
                </div>
            </section>
        </article>
    )
}