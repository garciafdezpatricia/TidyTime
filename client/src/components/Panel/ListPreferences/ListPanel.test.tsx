import { render } from "@testing-library/react";
import '@testing-library/jest-dom';
import ListPanel from '@/src/components/Panel/ListPreferences/ListPanel';

describe('ListPanel', () => {
    it('renders without crashing', () => {
      const { container } = render(<ListPanel />);
      expect(container).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<ListPanel />);
        expect(asFragment()).toMatchSnapshot();
      });
  });