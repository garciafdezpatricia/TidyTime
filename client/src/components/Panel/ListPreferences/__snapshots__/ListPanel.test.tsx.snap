// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPanel matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="list-preferences"
  >
    <h3>
      preferences.list.title
    </h3>
    <hr />
    <p>
      preferences.list.desc
    </p>
    <section
      class="default-labels"
    >
      <div
        style="display: flex; align-items: center; gap: .5rem;"
      >
        <p>
          Loading...
        </p>
        <div
          class="loader"
        />
      </div>
      <button
        class="new-label"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 12H3"
          />
          <path
            d="M16 6H3"
          />
          <path
            d="M16 18H3"
          />
          <path
            d="M18 9v6"
          />
          <path
            d="M21 12h-6"
          />
        </svg>
      </button>
    </section>
    <section
      class="show-tasks-in-calendar"
    >
      <p>
        preferences.list.showTasks
      </p>
      <div
        class="show-buttons"
      >
        <button
          class=""
        >
          preferences.list.showYes
        </button>
        <button
          class="active-button"
        >
          preferences.list.showNo
        </button>
      </div>
    </section>
  </article>
</DocumentFragment>
`;
